core/dagop.go

```diff
@@ -321,44 +339,384 @@ func (op RawDagOp) Exec(ctx context.Context, g bksession.Group, inputs []solver.
+		return nil, fmt.Errorf("expected FS to be selected, instead got %T", obj)
+	}
+}
+
+// getAllContainerMounts gets the list of all mounts for a container, as well as all the
+// inputs that are part of it. Each mount's Input maps to an index in the
+// returned states.
+func getAllContainerMounts(container *Container) (mounts []*pb.Mount, states []llb.State, _ error) {
+	outputIdx := 0
+	inputIdxs := map[digest.Digest]int{}
+
+	// addMount converts a ContainerMount and creates a corresponding buildkit
+	// mount, creating an input if required
+	addMount := func(mnt ContainerMount) error {
+		st, err := defToState(mnt.Source)
+		if err != nil {
+			return err
+		}
+
+		mount := &pb.Mount{
+			Dest:     mnt.Target,
+			Selector: mnt.SourcePath,
+			Output:   pb.OutputIndex(outputIdx),
+			// ContentCache: nil,
+		}
+		if st.Output() == nil {
+			mount.Input = pb.Empty
+		} else {
+			dgst := digest.FromBytes(mnt.Source.Def[len(mnt.Source.Def)-1])
+			if idx, ok := inputIdxs[dgst]; ok {
+				mount.Input = pb.InputIndex(idx)
+			} else {
+				// track and cache this input index, since duplicates are unnecessary
+				// also buildkit's FileOp (which is underlying our DagOp) will
+				// remove them if we don't, which results in significant confusion
+				mount.Input = pb.InputIndex(len(states))
+				inputIdxs[dgst] = len(states)
+				states = append(states, st)
+			}
+		}
+
+		if mount.Readonly {
+			mount.Output = pb.SkipOutput
+			mount.Readonly = true
+		}
+
+		if mnt.CacheVolumeID != "" {
+			mount.Output = pb.SkipOutput
+			mount.MountType = pb.MountType_CACHE
+			mount.CacheOpt = &pb.CacheOpt{
+				ID: mnt.CacheVolumeID,
+			}
+			switch mnt.CacheSharingMode {
+			case CacheSharingModeShared:
+				mount.CacheOpt.Sharing = pb.CacheSharingOpt_SHARED
+			case CacheSharingModePrivate:
+				mount.CacheOpt.Sharing = pb.CacheSharingOpt_PRIVATE
+			case CacheSharingModeLocked:
+				mount.CacheOpt.Sharing = pb.CacheSharingOpt_LOCKED
+			}
+		}
+
+		if mnt.Tmpfs {
+			mount.Output = pb.SkipOutput
+			mount.MountType = pb.MountType_TMPFS
+			mount.TmpfsOpt = &pb.TmpfsOpt{
+				Size_: int64(mnt.Size),
+			}
+		}
+
+		mounts = append(mounts, mount)
+		if mount.Output != pb.SkipOutput {
+			outputIdx++
+		}
+
+		return nil
+	}
+
+	// handle our normal mounts
+	if err := addMount(ContainerMount{Source: container.FS, Target: "/"}); err != nil {
+		return nil, nil, err
+	}
+	if err := addMount(ContainerMount{Source: container.Meta, Target: buildkit.MetaMountDestPath, SourcePath: buildkit.MetaMountDestPath}); err != nil {
+		return nil, nil, err
+	}
+	for _, mount := range container.Mounts {
+		if err := addMount(mount); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	// handle secret mounts
+	for _, secret := range container.Secrets {
+		if secret.MountPath == "" {
+			continue
+		}
+		uid, gid := 0, 0
+		if secret.Owner != nil {
+			uid, gid = secret.Owner.UID, secret.Owner.GID
+		}
+		mount := &pb.Mount{
```

Comment:
should this have `Output: pb.SkipOutput` for good measure if nothing else?

same comment for sockets below

Reply:
Mmmm, this didn't *seem* to be breaking anything, but yes, worth fixing anyways :scream: